libraries/chain/include/eosio/chain/fork_database.hpp

```diff
@@ -149,24 +151,26 @@ namespace eosio::chain {
       void close();
 
       // expected to be called from main thread
-      void switch_from_legacy(const block_handle& bh);
+      void switch_from_legacy(const block_state_variant_t& bhv);
 
       bool fork_db_if_present() const { return !!fork_db_s; }
       bool fork_db_legacy_present() const { return !!fork_db_l; }
 
       // see fork_database_t::fetch_branch(forkdb->head()->id())
       block_branch_t fetch_branch_from_head() const;
 
+      void reset_root(const block_state_variant_t& v);
+
       template <class R, class F>
       R apply(const F& f) const {
          if constexpr (std::is_same_v<void, R>) {
-            if (legacy) {
+            if (in_use.load(std::memory_order_relaxed) == in_use_t::legacy) {
```

Comment:
I don't think we can't use `memory_order_relaxed` here and below. We want to make sure access to `in_use` is ordered with access to `fork_db_l` and `fork_db_s`.

Reply:
OK, that doesn't buy us  much because, even with `std::memory_order_seq_cst`, another thread could interrupt us and modify `in_use` after we load the value.
If no other thread is modifying in-use, then the memory_order doesn't matter.